lib/jxl/version.h.in

```diff
@@ -0,0 +1,27 @@
+/* Copyright (c) the JPEG XL Project Authors. All rights reserved.
+ *
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file.
+ */
+
+/** @addtogroup libjxl_common
+ * @{
+ * @file version.h
+ * @brief libjxl version information
+ */
+
+#ifndef JXL_VERSION_H_
+#define JXL_VERSION_H_
+
+#define JPEGXL_LIBRARY_VERSION @JPEGXL_LIBRARY_VERSION@
+#define JPEGXL_MAJOR_VERSION @JPEGXL_MAJOR_VERSION@
```

Comment:
I'd recommend against doing this. This will tell you the version of the header file used at compile time, but the version used at runtime might be different (if they have the same SOVERSION).

If you still want this information at compile time for whatever library version was used at compile time, you can get it from the pkg-config in your project; but I wouldn't encourage people using libjxl to rely on MAJOR.MINOR.PATCH versions at compile time. You have them at runtime.

Reply:
That's true as well.

What would be nice though is to have a convenience runtime function that checks version against a desired triplet instead of doing 1000000 * maj + 1000 * min + patch all the time then (which is error-prone and less readable). Something like [this](https://github.com/Exiv2/exiv2/blob/fde8ed07b32fd99e9d7b9353c19d3e4805978306/src/version.cpp#L110-L113).

Reply:
You have them at runtime, but you first need to get your application compiled before there is a runtime. If you want to write code that can compile against various SOVERSIONs, having the version defined after including jxl/encode.h is certainly more convenient than having to make a complicated build script that ends up getting the SOVERSION via pkg-config and doing the define that way.

I agree that we do want to discourage getting the full version info at compile time, since the runtime version might indeed be different and that could lead to unexpected problems.

For things that require knowing the version at compile time (e.g. optionally using a new function that was introduced), I think it is useful to have it available at compile time; for things that don't require that (e.g. setting a parameter to a value that will cause a "not supported" error in an older version), it should be checked at runtime.

I propose instead of exposing the full MAJOR.MINOR.PATCH version, only the MAJOR and MINOR versions are exposed, which should basically be the same thing as the SOVERSION (unless I'm missing something?). If you want to know the PATCH version too, then you have to get it at runtime.

Does that make sense?

Reply:
> If you still want this information at compile time

as Jon pointed out, it is required to have that information at compile time, unless you are either bundling a known fixed version of jxl, or jxl guarantees it will never modify its public facing api.

> you can get it from the pkg-config in your project

doing so will result in the exact same behavior at runtime, but requires jumping through hoops to extract that information from pkg-config.

Reply:
> I propose instead of exposing the full MAJOR.MINOR.PATCH version, only the MAJOR and MINOR versions are exposed

I'm fine with that, I agree there should be no reason to know the patch version at compile time.

Reply:
> That's true as well.
> 
> ~What would be nice though is to have a convenience runtime function that checks version against a desired triplet instead of doing 1000000 * maj + 1000 * min + patch all the time then (which is error-prone and less readable). Something like [this](https://github.com/Exiv2/exiv2/blob/fde8ed07b32fd99e9d7b9353c19d3e4805978306/src/version.cpp#L110-L113).~
> 
> Scratch that, the `JPEGXL_COMPUTE_LIBRARY_VERSION` macro is enough (maybe just simplify to `JXL_COMPUTE_VERSION`)...

For the record, the intended usage is to be able to write stuff like
```c
#if JPEGXL_VERSION_NUM < JPEGXL_COMPUTE_LIBRARY_VERSION(0.7.0)
//use old api
#else
//use new api
#endif
```

Reply:
SOVERSION is not MAJOR.MINOR, is a different number in cmake. For versions before <1.0  we can change the soversion between minor updates, but not really after 1.0.

"optionally using a new function that was introduced" is actually not an example supporting having this. Optionally using a new function that was introduced means that your compiled program only works with that version of libjxl or newer ones that have the function too, but would give you undefined reference errors at runtime if you try to use an older minor patch even if it has the same soversion and is otherwise meant to be compatible. Having this macro with an ifdef in your code gives you a false impression of more compatibility with no actual benefit in terms of compatibility, it will bite you. What projects normally do instead is they update their dependency to say "need libjxl>0.7" and use the function unconditionally. If you want to use the older libjxl, then use the older version of the project. If you want to support both the new version and the old version of libjxl you can set that up in your project using the version from pkg-config, but then you also need to declare that you require certain version of libjxl when packaging your project based on what was used to compile. The ifdef doesn't solve it, and it only "works" unless you start dealing with distributing the code or packages.

Using the version from pkg-config in your project would be equivalent to have them already in a .h header like proposed here, yes, those are both compile-time. Arguably having it from pkg-config is better because you can also make config-time decisions about it (example, libjxl 0.7 perhaps includes a CMS api so you use that, otherwise you use your own implementation and link against lcms in your project/own implementation). However, the big difference is that by having you own project build time macro to decide if you use a new libjxl feature you are making that call explicitly in your project, and you have a place where to do that. Bugs due to improper handling of versioning in your project are your problem then. Having the version in an header won't make things easier for downstream projects, only appear to make them easier.

This is not a complicated build script thing, it is the standard way to declare your dependencies (like "libjxl >= 0.6", or `pkg_check_modules(JXL libjxl>=0.7)` when importing in cmake). If you want to obtain the version `pkg-config --modversion libjxl` will tell you the version; or in cmake the `<PREFIX>_VERSION` variable has it too. autotools has something similar.

Reply:
[so](https://github.com/webmproject/libwebp/blob/master/src/webp/encode.h#L23) [every](https://github.com/glennrp/libpng/blob/libpng16/png.h#L288) [other](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/c23672ce52ae53bd846b555439aa0a070b6d2c07/jconfig.h.in#L4) [library](https://github.com/mirrorer/giflib/blob/master/lib/gif_lib.h#L14) [is](https://github.com/uclouvain/openjpeg/blob/master/src/lib/openjp2/opj_config.h.cmake.in#L8) [doing](https://github.com/AOMediaCodec/libavif/blob/master/include/avif/avif.h#L58) [it](https://gitlab.com/libtiff/libtiff/-/blob/master/libtiff/tiffvers.h#L9) [wrong](https://github.com/FLIF-hub/FLIF/blob/master/src/library/flif.h#L26) ?

Reply:
@deymo I'm having a problem trying to comprehend what issue you are trying to prevent by objecting to export the version information. Whether client code is conditionally compiled depending on this version or not, it will not be robust against a libjxl dependency being upgraded or downgraded without a recompile; ensuring client code did not have access to the abi/api version at compile time, or was compiled with a very specific dependency version, will not protect you against that.

As exposed in the PR's initial example, knowing the api version at compile time is today required in order to be able to /compile/ the code against 0.6 or 0.7 libjxl versions. This mechanism has been used for ages to workaround quirks and capabilities when you don't have the hand on the exact dependency's version, which is a real use case in some environments.

Reply:
I don't really understand what you mean with protection "against" a libjxl upgraded/downgraded. We don't prevent people from upgrading their libraries. Upgrading libraries without recompiling all the code that depends on those libraries happens all the time (you don't re-ship every package when updating something basic used everywhere like libz), but when that happens it is done so that the new version is compatible with the old one. In those scenarios, the conditional compilation of optional features based on libjxl will not do what you expect by looking at the code, there will be users running new version of libjxl with your program compiled using the old api, no warnings issued.

The PRs example talks about a function that's deprecated, but not removed, it still works. Note that you say here that compiling against 0.7 gives you an error, but you don't have that error if you compile against 0.6 and run against 0.7 (assuming only the deprecation change), and you seem to be fine with that, even though it is the exact same situation at runtime. So my point is that relying on the installed version from the headers doesn't "solve" the problem in the description, it only appears to solve it, but the "problem" (calling the deprecated function) still exists in the field and it will be harder to understand and reproduce when the time comes.

0.7 has not been released yet, but instead if we assume that it is released and there's some other API/ABI incompatibilities between 0.6 and 0.7 (which are allowed under semantic versioning because of the major version is still 0) we can bump the SOVERSION to declare that incompatibility and force distributions shipping libjxl to recompile dependencies based on that because it is a different dependency file. This will prevent a code compiled against 0.6 library to run against 0.7 library because the SOVERSION is different (it also allows some use cases like to install both at the same time because the .so has a different name, this is more common when you have version 1.x and 2.x and have lots of projects that have not migrated to the new api, let's ignore this part). If the meaning of some function changes, or the size of a struct changes which is possible given that we are still building parts of the API, calling the old function is a big problem (libjxl and caller don't agree on the struct size), even if you rely on version.h numbers to decide whether to use a field or not. Failing to run (because of missing shared lib) is better than a segfault that goes away when you recompile your project to debug it. Does that make sense? So my point is that the ifdef with the version.h doesn't solve these problems and I don't want to encourage people to believe that it does.

We can have this, I just don't think is a good idea for the issues stated here. You can also get the information from the pkg-config if you know what you are doing WRT conditionally use dependency features based on compile-time version. pkg-config is not an obscure thing that nobody needs to look at. That's where we declare what flags you need to pass to compile/link against libjxl shared or static lib. The version is another metadata field there. Not using those flags will eventually lead to other problems too.

Reply:
> [so](https://github.com/webmproject/libwebp/blob/master/src/webp/encode.h#L23) [every](https://github.com/glennrp/libpng/blob/libpng16/png.h#L288) [other](https://github.com/libjpeg-turbo/libjpeg-turbo/blob/c23672ce52ae53bd846b555439aa0a070b6d2c07/jconfig.h.in#L4) [library](https://github.com/mirrorer/giflib/blob/master/lib/gif_lib.h#L14) [is](https://github.com/uclouvain/openjpeg/blob/master/src/lib/openjp2/opj_config.h.cmake.in#L8) [doing](https://github.com/AOMediaCodec/libavif/blob/master/include/avif/avif.h#L58) [it](https://gitlab.com/libtiff/libtiff/-/blob/master/libtiff/tiffvers.h#L9) [wrong](https://github.com/FLIF-hub/FLIF/blob/master/src/library/flif.h#L26) ?

@tbonfort First of all, perhaps the tone of this comment is not the most appropriate for a constructive discussion.

I didn't look into all of these, but I want to point out libjpeg case, which has a long history and had time to run into and sort out these issues, and is a project I had to deal with WRT versioning issues. In my Debian, that include file has `#define JPEG_LIB_VERSION  62`, which is also the SONAME of the installed libjpeg.so. So effectively if you build against 62 and use features from 62, your program won't run against other versions like 63. You can check this looking at what the binary file links against:
```
readelf -d /usr/bin/cjpeg | grep NEED
 0x0000000000000001 (NEEDED)             Shared library: [libjpeg.so.62]
```

I have libjpeg-turbo installed, so the version (major.minor.patch) from libjpeg-turbo is a different unrelated number. The comment in that file says to use this 62 number for ifdef checks. So, no, it doesn't look like jpeg got it wrong, yet this "62" version is not what your PR is proposing to add. Perhaps we should call this an "API level" number for more clear understanding. For pre-1.0 there can be incompatible changes; but for 1.0 and later API incompatibilities in our part require us to make a new major release (like 2.0).

Perhaps this makes it more clear.

Reply:
So I would suggest to put an API level define in the .h (for applications that want to be compilable with various API versions), and keep the 
version number info only accessible through the runtime function.

Reply:
@deymo thanks for taking the time to explain your point of view, and sorry for the aggressive/frustrated tone.

> The PRs example talks about a function that's deprecated, but not removed, it still works. Note that you say here that compiling against 0.7 gives you an error, but you don't have that error if you compile against 0.6 and run against 0.7

My objective is not to compile against one version and run against another, but solely to run and compile under different libjxl versions. The example I gave does cause an issue today as the calling code is compiled with -Werror -Wno-deprecated, but the heart of the issue still stands as you could have very well decided to delete the previous function completely given the current 0.x status (or were bumping the major version). Making sure the client code can compile cleanly both against 0.6 and 0.7 can only be addressed through conditional compilation, and as shown by my linked examples the standard way of determining that conditional is by reading the API version from an exported header (I understand that technically that "define" can be extracted at configure time from pkg-config, but that's not something I have often/ever seen, and the end result in terms of compiled code is identical).

>  yet this "62" version is not what your PR is proposing to add. Perhaps we should call this an "API level" number for more clear understanding

Yes, this may be a misunderstanding, because this "62" version is in essence the information I would like to add to jxl headers with this PR. I am somewhat confused about the difference between an ABI and API version when the exposed API/ABI is only in C and not C++. i.e. my understanding is that for a C API (which is the case for jxl), API and ABI can be used interchangeably.

Reply:
Correcting myself concerning C API/ABI forward compatibility: it is my understanding that there *only* can be a gap between a C library's API vs ABI minor version increment when you are adding a member to a publicly exposed struct. This case can however be overcome if you define your API to be
```c
typedef struct foobar {
  ...
}
foobar* initFoobar();
int library_function(foobar *fb);
void releaseFoobar(foobar *fb);
```
and then subsequently add new members to foobar at the end of the struct. With this mechanism in place, client code compiled against vx.y of the ABI will be forward compatible with vx.y+n versions when new members have been added to the struct and initFoobar has initialised the new member to something backwards compatible.